code/espurna/ota.ino

```diff
@@ -30,8 +30,22 @@ void _otaLoop() {
 
 #if TERMINAL_SUPPORT || OTA_MQTT_SUPPORT
 
-#include <ESPAsyncTCP.h>
-AsyncClient * _ota_client;
+#ifndef ARDUINO_ESP8266_RELEASE_2_5_0
+#define USING_AXTLS // do not use BearSSL
+#endif
+
+WiFiClient _ota_client;
```

Comment:
Using single _ota_client object instead of separate non-secure and secure ?
Minor question - does this create _ota_client when secure one is used? (i.e. is it in memory at this point?)

Reply:
I thought about it but it clutters the code a lot if you're using a separate _secure object: all subsequent ``_ota_client`` calls need to be surrounded with if-checks. But maybe this is not the nicest solution. Now that I think about it - not sure what happens when connecting to a HTTPS host first and to a HTTP one later. Do you have any suggestions?

Reply:
My bad not looking at this closer. Secure client will not fallback to basic one :(

Essentially, what my thinking is
- use separate methods for secure and non-secure update flow
- remove global client objects, create them inplace. We are not async, we expect them to live up until the function ends.

Reply:
And mqtt /ota handler needs to be aware of that too. All async-mqtt-client callbacks are inside SYS context - we cannot use Core's sync clients inside of it

Reply:
I just pushed some changes, the client objects are created inside the function now. I have a strange crash with SSL requests though (alloc issues in SSL) - can you check on your end as well?

Reply:
Created dummy server from dockered nginx:1.15-alpine and letsencrypt cert chain. Fingerprinted or not, it either gets stuck on connection or errors out. Not even http request logged, unlike opening same .bin with browser.
```
ota https://<server>/firmware.bin
[014623] +OK
[014624] [OTA] Connecting to <server>:443 to download /firmware.bin
[014625] [OTA] Using fingerprint: <bla bla bla>
ERROR[0]: No Error
[015533] [OTA] Disconnected
```

Reply:
Something that I tried yesterday with HTTPClient & httpUpdate was requiring ~20seconds for connection to happen.

In general, I think we would want to avoid doing connection twice. HTTPClient, for example, wraps around that internally (which btw can be adapted here or straight up used as-is)

And  I did rewrite bunch of things to stop axtls thing from happening. I think one problematic part is shadowing _ota_client instead of having separate methods. Reviewing HTTPClient I'd noticed that it cleverly wraps around that too (see Traits classes)

Reply:
Remedy for 20 second wait was turning off active webserver (that was not ok...)
```
if (getSetting("webEnable", 1).toInt() == 0) return;
```
In webSetup(), wsSetup() and debugWebSetup()

At least now I have a reproducible "conflict" mentioned in the comments (which I did not remember having earlier :/). I wonder if async webserver polling method is broken somehow

Reply:
In my opinion it would be better to remove ESPAsyncTCP/ESPAsyncWebServer in the whole project - it's not maintained anymore and you'll see more and more compatibility issues with future core versions. Rewriting the telnet server is not too much work (I'm testing it now). For web support: the standard ESP8266WebServer is pretty good, you'll only need something like arduinoWebSockets for WS(S) support.

Reply:
Likely thing :( Web server part is problematic for some parts though - alexa triggers header parsing bugs, some credentials stuff works differently... Which again puts the Core version as a dependency on some bug or feature. *(OT: we also can switch to espressif IDF 🙄)*
Websocket is not a huge problem, since it is both an [external lib and pretty widely used](https://github.com/Links2004/arduinoWebSockets)

Reply:
Another heads up regarding slightly getting distracted and dependencies in general.
What we can do instead of removing the module straight away, is implement another one side by side (just another file) and make them incompatible with each other. That way we can already use new one, and don't need to thing about dependencies in a strange way.